allow sorting keys on to_json and to_python by passing in sort_keys

aezomz opened the pull request:
<!-- Thank you for your contribution! -->
Hello Pydantic Team! This is my first time contributing to a Rust and Pyo3 related repo.
I am also new in Rust. 
Do you think this PR will make sense? 
Since I have been trying to do model_dump_json with sort keys too.

This feature should simulate the same as how we use `json.dumps(data, sort_keys=True)`
```
Will sort from:
        {
            'field_123': b'test_123',
            'field_b': 12,
            'field_a': b'test',
            'field_c': {'mango': 2, 'banana': 3, 'apple': 1},
            'field_n': [
                {'mango': 3, 'banana': 2, 'apple': 1},
                [{'mango': 3, 'banana': 2, 'apple': 1}, {'d': 3, 'b': 2, 'a': 1}],
                3,
            ],
            'field_d': [
                {'d': 3, 'b': 2, 'a': {'nested3': 3, 'nested1': 1, 'nested2': 2}},
                [[{'mango': 3, 'banana': 2, 'apple': 1}], {'d': 3, 'b': 2, 'a': 1}],
                3,
            ],
            'field_none': None,
        }
To:
    assert s.to_python(m, exclude_none=True, sort_keys=True) == snapshot(
        {
            'field_123': b'test_123',
            'field_a': b'test',
            'field_b': 12,
            'field_c': {'apple': 1, 'banana': 3, 'mango': 2},
            'field_n': [
                {'apple': 1, 'banana': 2, 'mango': 3},
                [{'apple': 1, 'banana': 2, 'mango': 3}, {'a': 1, 'b': 2, 'd': 3}],
                3,
            ],
            'field_d': [
                {'a': {'nested1': 1, 'nested2': 2, 'nested3': 3}, 'b': 2, 'd': 3},
                [[{'apple': 1, 'banana': 2, 'mango': 3}], {'a': 1, 'b': 2, 'd': 3}],
                3,
            ],
        }
    )
```
Take note that
1. `field_d` is extra and still manage to sort
2. sorting recursively for both defined schema and extras
3. sorting including dictionary in array or nested array array

Please let me know if I miss out any other features that `sort_keys=True` is suppose to do!

Thanks!
## Change Summary
allow sorting keys on to_json and to_python by passing in sort_keys

<!-- Please give a short summary of the changes. -->

## Related issue number
should fix https://github.com/pydantic/pydantic/issues/7424
Might need to create another MR on Python repo though, need to check.

<!-- Are there any issues opened that will be resolved by merging this change? -->
<!-- WARNING: please use "fix #123" style references so the issue is closed when this PR is merged. -->

## Checklist

* [x] Unit tests for the changes exist
* [x] Documentation reflects the changes where applicable
* [X] Pydantic tests pass with this `pydantic-core` (except for expected changes)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**


Selected Reviewer: @davidhewitt

aezomz commented:
please review

adriangb commented on src/errors/validation_exception.rs:
```diff
@@ -347,6 +347,7 @@ impl ValidationError {
             None,
             false,
             false,
+            false,
             true,
```
@davidhewitt this is getting gross 😢

adriangb commented on src/serializers/fields.rs:
```diff
@@ -156,9 +156,12 @@ impl GeneralFieldsSerializer {
-        // NOTE! we maintain the order of the input dict assuming that's right
-        for result in main_iter {
-            let (key, value) = result?;
+        let mut items = main_iter.collect::<PyResult<Vec<_>>>()?;
```
My intuition is that this extra Vec being created is the cause for the slowdown in benchmarks.

aezomz replied:
I refactored the function out, so we can re-use this. and avoid creating new variable when sort keys is false

adriangb commented on tests/serializers/test_model.py:
```diff
@@ -991,22 +992,22 @@ def test_extra():
         MyModel,
         core_schema.model_fields_schema(
             {
-                'field_a': core_schema.model_field(core_schema.bytes_schema()),
                 'field_b': core_schema.model_field(core_schema.int_schema()),
+                'field_a': core_schema.model_field(core_schema.bytes_schema()),
```
Why did these have to change? Is this to make the tests change? I would prefer a new self-contained test just for this bit.

aezomz replied:
seperated this.

adriangb reviewed (commented):
Main issue is perf regression

aezomz commented:
I separated the test out, I also refactor the functions so we can reuse when `sort_keys=true`. 
To keep the original perf benchmark, I have done a simple bool check on sort_keys before using expensive function like sorting.

Let me know what else I need to improve.  Thanks

adriangb reviewed (commented):
Hmm I see now that this is not recursive (it only applies to the top level keys). Would it be hard to make it recursive? I fear that if we implement the non-recursive version someone is going to come along and want the recursive version... if so we can make it a `Literal['recursive', 'top-level', 'unsorted']` or something like that.

aezomz commented:
> Hmm I see now that this is not recursive (it only applies to the top level keys). Would it be hard to make it recursive? I fear that if we implement the non-recursive version someone is going to come along and want the recursive version... if so we can make it a `Literal['recursive', 'top-level', 'unsorted']` or something like that.

Added different sort mode as above, updated the PR description.

aezomz commented on src/serializers/fields.rs:
```diff
@@ -156,56 +156,28 @@ impl GeneralFieldsSerializer {
-
-                    if field.required {
+        } else {
+            let mut items = main_iter.collect::<PyResult<Vec<_>>>()?;
+            items.sort_by_cached_key(|(key, _)| key_str(key).unwrap_or_default().to_string());
```
Sorry, I kept the original implementation. Didn't quite get your advice, still quite new to Rust 😓

adriangb replied:
The first thing `process_field_entry_python` does is call `key_str`: https://github.com/pydantic/pydantic-core/blob/95f9329bff9171886f2d7bad99c424391357abfc/src/serializers/fields.rs#L234

aezomz replied:
I updated to convert it once and pass to the function.

aezomz commented:
please review 👍

aezomz commented:
please review 👍

samuelcolvin commented on python/pydantic_core/_pydantic_core.pyi:
```diff
@@ -305,6 +305,7 @@ class SchemaSerializer:
         exclude_unset: bool = False,
         exclude_defaults: bool = False,
         exclude_none: bool = False,
+        sort_keys: Literal['recursive', 'top-level', 'unsorted'] = 'unsorted',
```
do we really need these three options? surely if we add this feature it should match python's `json.dumps` behaviour and be `sort_keys: bool = False`.

adriangb replied:
There are scenarios where you'd want to only sort the top level, e.g. the trick we were doing with our JSON data in FusionFire.

But I also agree that if no one is asking for it we shouldn't implement extra stuff.

How does it sound if for now we do `sort_keys: bool = False` and then if needed in the future we can do `sort_keys: bool | Literal['top-level'] = False`.
We reduce the complexity and make a nicer signature for now but can still add the top-level option in the future, even if it makes the signature a bit weird.

@samuelcolvin @aezomz wdyt?

samuelcolvin replied:
agreed, that's what we've done for previous flags.

aezomz replied:
Sure, let me check the feature for sort_keys on json.dumps.
I only want to do `sort_keys` on `model_dump_json` thats why am here 😆 
Let me try to implement this.

adriangb replied:
Thank you so much @aezomz, you are being incredibly patient and helpful 🙏🏻

aezomz replied:
Hi Adrian, please review when ur free. I think David fixed the wasm build test already.

aezomz committed:
feat: allow sort_keys on to_json and to_python

aezomz commented:
please review 👍

There is this test that is failing, unclear if its related to my change.
[build-wasm-emscripten](https://github.com/pydantic/pydantic-core/actions/runs/14200249977/job/39785205616?pr=1637#logs)

Let me know any other places i can optimize. Thanks!

aezomz commented:
please review 👍

aezomz commented on src/serializers/fields.rs:
```diff
@@ -229,6 +208,146 @@ impl GeneralFieldsSerializer {
         }
     }
 
+    fn contains_dict(value: &Bound<'_, PyAny>) -> bool {
+        if value.downcast::<PyDict>().is_ok() {
+            return true;
+        }
+        if let Ok(list) = value.downcast::<PyList>() {
+            for item in list.iter() {
+                if item.downcast::<PyDict>().is_ok() {
+                    return true;
+                }
+            }
+        }
+        false
+    }
```
it will fail this test code currently:
 ```
m = v.validate_python(
        {
            'field_123': b'test_123',
            'field_b': 12,
            'field_a': b'test',
            'field_c': {'mango': 2, 'banana': 3, 'apple': 1},
            'field_n': [{'mango': 3, 'banana': 2, 'apple': 1}, 2, 3],
        }
    )

assert s.to_json(m, sort_keys=True) == snapshot(
        b'{"field_123":"test_123","field_a":"test","field_b":12,"field_c":{"apple":1,"banana":3,"mango":2},"field_n":[{"apple":1,"banana":2,"mango":3},2,3]}'
    )

----
E         Full diff:
E           (b'{"field_123":"test_123","field_a":"test","field_b":12,"field_c":{"apple":1,"'
E         -  b'banana":3,"mango":2},"field_n":[{"apple":1,"banana":2,"mango":3},2,3]}')
E         ?                                       ^^^^  ^             - ^^^  ^
E         +  b'banana":3,"mango":2},"field_n":[{"mango":3,"banana":2,"apple":1},2,3]}')
E         ?   
```

adriangb replied:
Hmm I can't tell why this is.

It just seems to me that if we have `list -> dict -> list -> dict` etc. we shouldn't have to peek into the lists to see if they have dicts, we have to just recurse each item once. Is there something I'm missing or do you agree?

aezomz replied:
Did u manage to look at this ?
https://github.com/pydantic/pydantic-core/pull/1637#discussion_r2042624300

Are u trying to say that i shouldnt peek first and do the below irregardless?
`let sorted_dict = Self::sort_dict_recursive(value.py(), value).map_err(py_err_se_err)?;`

aezomz replied:
@adriangb  do you mind taking a look ? thanks!

aezomz commented on src/serializers/fields.rs:
```diff
@@ -242,24 +361,67 @@ impl GeneralFieldsSerializer {
+            if let Some(field) = self.fields.get(key_str) {
+                if let Some(ref serializer) = field.serializer {
+                    if !exclude_default(value, &field_extra, serializer).map_err(py_err_se_err)? {
+                        if extra.sort_keys && Self::contains_dict(value) {
```
@adriangb , we use `contains_dict` check here as well. This is to ensure to check for any dictionary before we commit to actually try to sort.

aezomz commented:
please review 👍

aezomz commented:
please review 👍

aezomz commented:
@DouweM please review :)

DouweM commented on src/serializers/fields.rs:
```diff
@@ -229,6 +208,124 @@ impl GeneralFieldsSerializer {
         }
     }
 
+    fn sort_dict_recursive<'py>(py: Python<'py>, value: &Bound<'py, PyAny>) -> PyResult<Bound<'py, PyAny>> {
```
I know this was suggested in a previous review, but instead of doing this recursively here, would it be an option to implement single-level sorting in this file _and_ in `src/serializers/type_serializers/dict.rs`, so that it works on any dict serialization, not just when nested in a model/dataclass/typeddict?

That way, we may not need this recursive helper function anymore, and possibly can do away with `process_field_entry_python` entirely, since `items.sort_by(|(a, _, _), (b, _, _)| a.cmp(b))` is already happening in `main_serde_serialize` and `main_to_python`.

aezomz replied:
Thanks for guiding me.
I did a refactor based on your advice!
However, i still have to add a few functions to ensure that we dont repeat our code too much.

DouweM commented on src/serializers/fields.rs:
```diff
@@ -242,24 +339,67 @@ impl GeneralFieldsSerializer {
         // we don't both with `used_fields` here because on unions, `to_python(..., mode='json')` is used
```
The comment just above here says "we maintain the order of the input dict", that's not correct anymore!

DouweM commented on src/serializers/fields.rs:
```diff
@@ -156,56 +156,35 @@ impl GeneralFieldsSerializer {
-            if extra.exclude_none && value.is_none() {
-                if let Some(field) = op_field {
-                    if field.required {
+        if !extra.sort_keys {
```
Can we swap these branches so we have the true case first followed by the false case? Same for the other instance of `if !extra.sort_keys`

DouweM reviewed (changes requested):
Thanks for your work on this @aezomz! I'm a Rust newbie as well, but I have a feeling we can significantly simplify this code by dropping the recursive dict sorting and moving this to the serializer for the dict type specifically. Can you give that a try please?

I also think it's worth seeing if we can reduce the duplication between the `if sort_keys`/`else` branches, but I know Rust typing may make that tricky...

Let me know if you get stuck, we can bring in some of our Rust experts!

DouweM commented on src/serializers/fields.rs:
```diff
@@ -238,28 +283,75 @@ impl GeneralFieldsSerializer {
+                    if !exclude_default(value, &field_extra, serializer).map_err(py_err_se_err)? {
+                        // Get potentially sorted value
+                        if extra.sort_keys {
+                            let sorted_dict = sort_dict_recursive(value.py(), value).map_err(py_err_se_err)?;
```
Why do we still need to `sort_dict_recursive` here if we're already doing that in the dict serialization itself? (Same question for the 2 calls below)

aezomz replied:
I try ur recommmendation and it doesn't work.

It will only work when converting to Python type (`to_python`). Since we wrote the serializing logic.
As for `to_json` we use the serde package if am not wrong.
They are separate into defined field processing and non defined fields processing. Cause we need to sort before using the serde package. 

This is based on my high level understanding.

DouweM commented on src/serializers/type_serializers/dict.rs:
```diff
@@ -92,8 +117,17 @@ impl TypeSerializer for DictSerializer {
-                        let value =
-                            value_serializer.to_python(&value, next_include.as_ref(), next_exclude.as_ref(), extra)?;
+                        let value = if extra.sort_keys {
+                            let sorted_value = sort_dict_recursive(py, &value)?;
```
Could we not do a recursive sort here, but sort `new_dict` specifically after we've built it here?

DouweM commented on src/serializers/type_serializers/dict.rs:
```diff
@@ -128,14 +162,26 @@ impl TypeSerializer for DictSerializer {
-                        );
-                        map.serialize_entry(&key, &value_serialize)?;
+                        if extra.sort_keys {
+                            let sorted_dict = sort_dict_recursive(value.py(), &value).map_err(py_err_se_err)?;
```
Same as above -- I'd like to get rid of recursive sorting and do this after we build the `map` here. But I'm not 100% sure the data types here allow that...

DouweM reviewed (changes requested):
@aezomz Thanks, this looks a lot better already, but I have a feeling we can get rid of the recursive sorting entirely, if we do it at each level (model + dict) explicitly.

I'm running into the limits of my Rust skills though, and I'm not sure the data types involved actually allow us to do this.

@davidhewitt Would you mind having a Rusty eyed look for us? :)

DouweM reviewed (changes requested):
@aezomz Thanks, this looks a lot better already, but I have a feeling we can get rid of the recursive sorting entirely, if we do it at each level (model + dict) explicitly.

I'm running into the limits of my Rust skills though, and I'm not sure the data types involved actually allow us to do this.

@davidhewitt Would you mind having a Rusty eyed look for us? :)

aezomz commented on src/serializers/fields.rs:
```diff
@@ -238,28 +283,75 @@ impl GeneralFieldsSerializer {
+            ..*extra
+        };
+
+        let filter = self.filter.key_filter(key, include, exclude).map_err(py_err_se_err)?;
```
This check for include and exclude defined in the parameters. Those fields are definied in pydantic schema

aezomz commented on src/serializers/fields.rs:
```diff
@@ -444,8 +546,19 @@ impl TypeSerializer for GeneralFieldsSerializer {
                     let output_key = infer_json_key(&key, extra).map_err(py_err_se_err)?;
-                    let s = SerializeInfer::new(&value, next_include.as_ref(), next_exclude.as_ref(), extra);
-                    map.serialize_entry(&output_key, &s)?;
+                    if extra.sort_keys {
```
While this block is cater for extra fields (not defined in pydantic schema)
So we need to sort recursively for that as well. 
Both of them doesn't seem to use dict.rs serializer . As the objective is `to_json` and using serde package directly...

https://github.com/pydantic/pydantic-core/pull/1637/files#diff-d32436e9ac9b3e5dfaf920749269f6cff3dae3b3e030561f9b9bf50447067450R540

aezomz commented:
@DouweM  can you help to review again?
I didn't make much changes except for rebase and commit.
I have shared my comments above on the concerns.
Thanks!

davidhewitt reviewed (changes requested):
Sorry for the very slow review by me. Thank you for working on this.

I think this is a good starting point however there's a fair bit of work still needed to make this implementation both efficient and correct.

aezomz commented on src/serializers/infer.rs:
Oic, i just realized that the fields.rs json serialization uses infer.rs ... thats a great hint. Managed to implement it. Will look at adding the tests... Thanks David.

aezomz committed:
rebase and adapt to new compute field fn

aezomz commented:
@davidhewitt  please give it another review when u have the chance. thanks~

davidhewitt reviewed (commented):
Sorry for the very long review cycle here.

I think the implementation is now looking better, however I would still like to see less duplication, which I noted in a bunch of places.

It looks like we're also missing tests for various non-model types:
- the `infer.rs` stuff would want tests, probably in `test_any.py`
- we probably also need to test `dataclass`, `dict`, and `TypedDict` containers
